examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+## Example for calculating and plotting the Probability of Improvement
+
+# This example demonstrates how to use the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility
+# to calculate and analyse the probability of improvement (PI) evolving during iterations:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+import os
+import warnings
+from pathlib import Path
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.axes import Axes
+from matplotlib.collections import PolyCollection
+from matplotlib.figure import Figure
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+
+### Parameters for a full simulation loop
+
+# For the full simulation, we need to define some additional parameters. These are the number of experiments to be conducted per run, the batch size, the dimension and the points per dimension.
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 3 if SMOKE_TEST else 7
+BATCH_SIZE = 2
+DIMENSION = 3
+POINTS_PER_DIM = 4
+
+### Setup
+
+# We start by initializing the global `poi_list` where we will store the PI from each recommend iteration:
+
+pi_list: list[torch.Tensor] = []
+
+
+# Then, we define the hook that calculates the PI based on the given recommender, search space and measurements:
+
+
+def check_probability_of_improvement(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+) -> list[torch.Tensor]:
+    """Calculate and store the probability of improvement in poi_list.
+
+    For reasons of numerical stability, the function adds some noise to the calculated
```

Comment:
Can you explain to me why this is necessary?

Reply:
Cause how else do you want to explain adding a (very) small value to the pi?

Reply:
I'm not asking why you do it this way but why is it necessary in the first place 🙃 What would happen if you didn't add the small values, i.e. what kind of "numerical stability" are we talking about? Which operation would break?

Reply:
Oh sorry I misunderstood your point.  It is because we had a weird error caused by the `gaussian_kde()` when the `batch_size = 1`. 
Actually, I can take it out, because now we decided to have `batch_size = 2`.  Or should I leave it for more security?
@AVHopp What do you think?

Reply:
The problem was that the guassian_kde calculation fails when the PI is equal to 0.5 everywhere with the following error:

numpy.linalg.LinAlgError: The data appears to lie in a lower-dimensional subspace of the space in which it is expressed. This has resulted in a singular data covariance matrix, which cannot be treated using the algorithms implemented in gaussian_kde. Consider performing principle component analysis / dimensionality reduction and using gaussian_kde with the transformed data.

But this only happens for batch_size=1. We could say that we add the noise conditionally or remove it completely if we want to keep batch_size=2

Reply:
Since we already decided to keep batch_size = 2, I will just remove it + to make the example less complicated.